privacyidea/lib/tokens/pushtoken.py

```diff
@@ -0,0 +1,578 @@
+        for k, v in param.items():
+            upd_param[k] = v
+
+        if "serial" in upd_param and "fbtoken" in upd_param and "pubkey" in upd_param:
+            # We are in step 2:
+            if not self.token.rollout_state == "clientwait":
+                raise ParameterError("Invalid state! The token you want to enroll is not in the state 'clientwait'.")
+            if upd_param.get("enrollment_credential") != self.get_tokeninfo("enrollment_credential"):
+                raise ParameterError("Invalid enrollment credential. You are not authorized to finalize this token.")
+            self.del_tokeninfo("enrollment_credential")
+            self.token.rollout_state = "enrolled"
+            self.token.active = True
+            self.add_tokeninfo(PUBLIC_KEY_SMARTPHONE, upd_param.get("pubkey"))
+            self.add_tokeninfo("firebase_token", upd_param.get("fbtoken"))
+            # create a keypair for the server side.
+            from privacyidea.lib.crypto import generate_keypair
+            pub_key, priv_key = generate_keypair(4096)
+            #pub_key = _strip_key(pub_key)
+            #priv_key = _strip_key(priv_key)
+            self.add_tokeninfo(PUBLIC_KEY_SERVER, pub_key)
+            self.add_tokeninfo(PRIVATE_KEY_SERVER, priv_key, "password")
+
+        elif "genkey" in upd_param:
+            # We are in step 1:
+            upd_param["2stepinit"] = 1
+            self.add_tokeninfo("enrollment_credential", geturandom(20, hex=True))
+            # We also store the firebase config, that was used during the enrollment.
+            self.add_tokeninfo(PUSH_ACTION.FIREBASE_CONFIG, param.get(PUSH_ACTION.FIREBASE_CONFIG))
+        else:
+            raise ParameterError("Invalid Parameters. Either provide (genkey) or (serial, fbtoken, pubkey).")
+
+        TokenClass.update(self, upd_param, reset_failcount)
+
+    @log_with(log)
+    def get_init_detail(self, params=None, user=None):
+        """
+        This returns the init details during enrollment.
+
+        In the 1st step the QR Code is returned.
+        """
+        response_detail = TokenClass.get_init_detail(self, params, user)
+        if "otpkey" in response_detail:
+            del response_detail["otpkey"]
+        params = params or {}
+        user = user or User()
+        tokenlabel = params.get("tokenlabel", "<s>")
+        tokenissuer = params.get("tokenissuer", "privacyIDEA")
+        # Add rollout state the response
+        response_detail['rollout_state'] = self.token.rollout_state
+
+        extra_data = {"enrollment_credential": self.get_tokeninfo("enrollment_credential")}
+        imageurl = params.get("appimageurl")
+        if imageurl:
+            extra_data.update({"image": imageurl})
+        if self.token.rollout_state == "clientwait":
+            # Get the values from the configured PUSH config
+            fb_identifier = params.get(PUSH_ACTION.FIREBASE_CONFIG)
+            firebase_configs = get_smsgateway(identifier=fb_identifier)
```

Comment:
As ``get_smsgateway`` also searches for regular SMS gateways, should we add a filter for firebase gateways here?

Reply:
will do.

Reply:
done in 9036c0d7